src/stream/decoder.rs

```diff
@@ -125,14 +152,109 @@ impl<R: Read> Decoder<R> {
         unsafe {
             self.buffer.set_len(read);
         }
+
+        self.offset = 0;
         in_buffer.pos = 0;
         in_buffer.size = read;
-        // So we can't read anything: input is exhausted.
 
+        // So we can't read anything: input is exhausted.
         Ok(read > 0)
     }
 }
 
+impl<R: Read> Decoder<R> {
+    /// This function handles buffer_refill state of the read operation
+    /// It returns true if read operation should be stopped and false otherwise
+    fn handle_refill(&mut self, hint: RefillBufferHint, 
+        in_buffer: &mut zstd_sys::ZSTD_inBuffer, out_buffer: &mut zstd_sys::ZSTD_outBuffer)-> Result<bool, io::Error> {
+
+        let refilled = match self.refill_buffer(in_buffer) {
+            Err(ref err) if out_buffer.pos > 0 && err.kind() == io::ErrorKind::WouldBlock => {
+                // underlying reader was blocked but we've already put some data into the output buffer
+                // we need to stop this read operation so data won' be lost
```

Comment:
it will interrupt the read similar to current behavior. As an option we can have a loop and retry similar to what `write_from_offset` is doing today.